File: src/lib/calc.ts
// Pretty Calc — math expression engine.
// Tokenizer -> shunting-yard (RPN) -> evaluator. No eval(). Safe + predictable.

export type AngleMode = "DEG" | "RAD";

type Token =
	| { type: "num"; value: number }
	| { type: "op"; value: OpName }
	| { type: "func"; value: string }
	| { type: "paren"; value: "(" | ")" }
	| { type: "postfix"; value: "!" | "%" }
	| { type: "comma" };

type OpName =
	| "+"
	| "-"
	| "*"
	| "/"
	| "^"
	| "mod"
	| "neg" // unary minus
	| "mul"; // implicit multiplication

const FUNCTIONS = new Set([
	"sin",
	"cos",
	"tan",
	"asin",
	"acos",
	"atan",
	"sinh",
	"cosh",
	"tanh",
	"asinh",
	"acosh",
	"atanh",
	"ln",
	"log",
	"log2",
	"sqrt",
	"cbrt",
	"exp",
	"abs",
	"floor",
	"ceil",
	"round",
	"sign",
]);

const PRECEDENCE: Record<OpName, number> = {
	"^": 5,
	neg: 5,
	mul: 3,
	"*": 3,
	"/": 3,
	mod: 3,
	"+": 2,
	"-": 2,
};

const RIGHT_ASSOC = new Set<OpName>(["^", "neg"]);

const OP_ARITY: Record<OpName, number> = {
	"^": 2,
	"*": 2,
	"/": 2,
	mod: 2,
	"+": 2,
	"-": 2,
	neg: 1,
	mul: 2,
};

/** Build the set of function names so the UI can know what exists. */
export const FUNC_NAMES = [...FUNCTIONS];

function isValueEnd(t: Token | undefined): boolean {
	if (!t) return false;
	if (t.type === "num") return true;
	if (t.type === "paren" && t.value === ")") return true;
	if (t.type === "postfix") return true;
	return false;
}

function isValueStart(t: Token): boolean {
	return (
		t.type === "num" ||
		t.type === "func" ||
		(t.type === "paren" && t.value === "(")
	);
}

/** Parse a math expression string into tokens, resolving constants. */
function tokenize(input: string, constants: Record<string, number>): Token[] {
	const s = input.replace(/\s+/g, "");
	const tokens: Token[] = [];
	const pushImplicitMulIfNeeded = (next: Token) => {
		if (isValueStart(next) && isValueEnd(tokens[tokens.length - 1])) {
			tokens.push({ type: "op", value: "mul" });
		}
	};

	let i = 0;
	while (i < s.length) {
		const c = s[i];

		// Number literal: digits + dot.
		if ((c >= "0" && c <= "9") || c === ".") {
			let j = i + 1;
			let dot = c === ".";
			while (j < s.length) {
				const d = s[j];
				if (d >= "0" && d <= "9") {
					j++;
				} else if (d === "." && !dot) {
					dot = true;
					j++;
				} else {
					break;
				}
			}
			const num = Number.parseFloat(s.slice(i, j));
			if (Number.isNaN(num)) throw new CalcError("Malformed number");
			const next: Token = { type: "num", value: num };
			pushImplicitMulIfNeeded(next);
			tokens.push(next);
			i = j;
			continue;
		}

		// Identifier: function name or constant.
		if ((c >= "a" && c <= "z") || (c >= "A" && c <= "Z") || c === "_") {
			let j = i + 1;
			while (j < s.length && /[a-zA-Z0-9_]/.test(s[j])) j++;
			const name = s.slice(i, j).toLowerCase();
			let next: Token;
			if (name === "pi" || name === "e" || name === "ans") {
				if (!(name in constants)) {
					throw new CalcError(`Unknown constant: ${name}`);
				}
				next = { type: "num", value: constants[name] };
			} else if (name === "mod") {
				next = { type: "op", value: "mod" as OpName };
			} else if (name === "ln") {
				next = { type: "func", value: "ln" };
			} else if (FUNCTIONS.has(name)) {
				next = { type: "func", value: name };
			} else {
				throw new CalcError(`Unknown name: ${name}`);
			}
			pushImplicitMulIfNeeded(next);
			tokens.push(next);
			i = j;
			continue;
		}

		// Parentheses.
		if (c === "(" || c === ")") {
			const next: Token = { type: "paren", value: c };
			if (c === "(") pushImplicitMulIfNeeded(next);
			tokens.push(next);
			i++;
			continue;
		}

		if (c === ",") {
			tokens.push({ type: "comma" });
			i++;
			continue;
		}

		if (c === "!") {
			tokens.push({ type: "postfix", value: "!" });
			i++;
			continue;
		}

		if (c === "%") {
			tokens.push({ type: "postfix", value: "%" });
			i++;
			continue;
		}

		if (c === "+" || c === "-") {
			const prev = tokens[tokens.length - 1];
			const isUnary =
				!prev ||
				prev.type === "op" ||
				prev.type === "func" ||
				prev.type === "comma" ||
				(prev.type === "paren" && prev.value === "(");
			if (isUnary) {
				if (c === "-") tokens.push({ type: "op", value: "neg" });
				// unary plus is a no-op.
			} else {
				tokens.push({ type: "op", value: c });
			}
			i++;
			continue;
		}

		if (c === "*" || c === "×" || c === "·" || c === "\u2217") {
			tokens.push({ type: "op", value: "*" });
			i++;
			continue;
		}
		if (c === "/" || c === "\u00f7") {
			tokens.push({ type: "op", value: "/" });
			i++;
			continue;
		}
		if (c === "^" || c === "\u2211" /* not real exponent */) {
			tokens.push({ type: "op", value: "^" });
			i++;
			continue;
		}

		throw new CalcError(`Unexpected character: ${c}`);
	}

	return tokens;
}

/** Shunting-yard: infix token list -> RPN. */
function toRPN(tokens: Token[]): Token[] {
	const output: Token[] = [];
	const stack: Token[] = [];

	for (const t of tokens) {
		if (t.type === "num") {
			output.push(t);
		} else if (t.type === "func") {
			stack.push(t);
		} else if (t.type === "comma") {
			while (
				stack.length &&
				!(stack[stack.length - 1].type === "paren")
			) {
				output.push(stack.pop() as Token);
			}
			if (!stack.length) throw new CalcError("Misplaced comma");
		} else if (t.type === "op") {
			while (stack.length) {
				const top = stack[stack.length - 1];
				if (top.type === "func") {
					output.push(stack.pop() as Token);
					continue;
				}
				if (top.type === "op") {
					const topPrec = PRECEDENCE[top.value as OpName];
					const curPrec = PRECEDENCE[t.value];
					const rightAssoc = RIGHT_ASSOC.has(t.value);
					if (
						(rightAssoc && topPrec > curPrec) ||
						(!rightAssoc && topPrec >= curPrec)
					) {
						output.push(stack.pop() as Token);
						continue;
					}
				}
				break;
			}
			stack.push(t);
		} else if (t.type === "paren" && t.value === "(") {
			stack.push(t);
		} else if (t.type === "paren" && t.value === ")") {
			while (
				stack.length &&
				!(stack[stack.length - 1].type === "paren")
			) {
				output.push(stack.pop() as Token);
			}
			if (!stack.length) throw new CalcError("Mismatched parentheses");
			stack.pop(); // remove "("
			// Bring function to output.
			if (stack.length && stack[stack.length - 1].type === "func") {
				output.push(stack.pop() as Token);
			}
		} else if (t.type === "postfix") {
			output.push(t);
		}
	}

	while (stack.length) {
		const t = stack.pop() as Token;
		if (t.type === "paren") throw new CalcError("Mismatched parentheses");
		output.push(t);
	}

	return output;
}

function factorial(n: number): number {
	if (n < 0 || Math.floor(n) !== n) {
		throw new CalcError("Factorial requires non-negative integer");
	}
	if (n > 170) throw new CalcError("Factorial too large");
	let r = 1;
	for (let k = 2; k <= n; k++) r *= k;
	return r;
}

function applyFunc(name: string, x: number, angle: AngleMode): number {
	const toRad = (deg: number) =>
		angle === "DEG" ? (deg * Math.PI) / 180 : deg;
	const fromRad = (rad: number) =>
		angle === "DEG" ? (rad * 180) / Math.PI : rad;
	switch (name) {
		case "sin":
			return Math.sin(toRad(x));
		case "cos":
			return Math.cos(toRad(x));
		case "tan":
			return Math.tan(toRad(x));
		case "asin":
			return fromRad(Math.asin(x));
		case "acos":
			return fromRad(Math.acos(x));
		case "atan":
			return fromRad(Math.atan(x));
		case "sinh":
			return Math.sinh(x);
		case "cosh":
			return Math.cosh(x);
		case "tanh":
			return Math.tanh(x);
		case "asinh":
			return Math.asinh(x);
		case "acosh":
			return Math.acosh(x);
		case "atanh":
			return Math.atanh(x);
		case "ln":
			return Math.log(x);
		case "log":
			return Math.log10(x);
		case "log2":
			return Math.log2(x);
		case "sqrt":
			return Math.sqrt(x);
		case "cbrt":
			return Math.cbrt(x);
		case "exp":
			return Math.exp(x);
		case "abs":
			return Math.abs(x);
		case "floor":
			return Math.floor(x);
		case "ceil":
			return Math.ceil(x);
		case "round":
			return Math.round(x);
		case "sign":
			return Math.sign(x);
		default:
			throw new CalcError(`Unknown function: ${name}`);
	}
}

/** Evaluate an RPN token stream. */
function evaluateRPN(rpn: Token[], angle: AngleMode): number {
	const stack: number[] = [];
	for (const t of rpn) {
		if (t.type === "num") {
			stack.push(t.value);
		} else if (t.type === "postfix") {
			if (!stack.length) throw new CalcError("Syntax error");
			const a = stack.pop() as number;
			stack.push(t.value === "!" ? factorial(a) : a / 100);
		} else if (t.type === "op") {
			const arity = OP_ARITY[t.value];
			if (arity === 1) {
				if (!stack.length) throw new CalcError("Syntax error");
				const a = stack.pop() as number;
				if (t.value === "neg") stack.push(-a);
			} else if (arity === 2) {
				if (stack.length < 2) throw new CalcError("Syntax error");
				const b = stack.pop() as number;
				const a = stack.pop() as number;
				switch (t.value) {
					case "+":
						stack.push(a + b);
						break;
					case "-":
						stack.push(a - b);
						break;
					case "*":
					case "mul":
						stack.push(a * b);
						break;
					case "/":
						if (b === 0) throw new CalcError("Division by zero");
						stack.push(a / b);
						break;
					case "^":
						stack.push(a ** b);
						break;
					case "mod":
						if (b === 0) throw new CalcError("Mod by zero");
						stack.push(((a % b) + b) % b);
						break;
				}
			}
		} else if (t.type === "func") {
			if (!stack.length)
				throw new CalcError(`Missing argument for ${t.value}`);
			const a = stack.pop() as number;
			stack.push(applyFunc(t.value, a, angle));
		} else {
			throw new CalcError("Malformed expression");
		}
	}
	if (stack.length !== 1) {
		throw new CalcError(
			stack.length === 0 ? "Empty input" : "Syntax error",
		);
	}
	return stack[0] as number;
}

export class CalcError extends Error {
	constructor(msg: string) {
		super(msg);
		this.name = "CalcError";
	}
}

export type EvalOptions = {
	angleMode?: AngleMode;
	ans?: number;
};

/** Evaluate a math expression string to a number. Throws CalcError on bad input. */
export function evaluate(input: string, opts: EvalOptions = {}): number {
	const angle: AngleMode = opts.angleMode ?? "DEG";
	const ans = opts.ans ?? 0;
	const trimmed = input.trim();
	if (!trimmed) throw new CalcError("Empty input");
	const tokens = tokenize(trimmed, { pi: Math.PI, e: Math.E, ans });
	const rpn = toRPN(tokens);
	// quick check: ensure there's at least one value-producing token
	const valueless =
		tokens.length === 0 ||
		tokens.every(
			(t) =>
				t.type === "op" ||
				t.type === "comma" ||
				(t.type === "paren" && t.value === "("),
		);
	if (valueless) throw new CalcError("Syntax error");
	return evaluateRPN(rpn, angle);
}

/** Try to evaluate; return null on any error (used for live preview). */
export function tryEvaluate(input: string, opts: EvalOptions): number | null {
	try {
		const r = evaluate(input, opts);
		if (!Number.isFinite(r)) return null;
		return r;
	} catch {
		return null;
	}
}

/** Format a number for display: compact, rounds float fuzz, uses scientific for extremes. */
export function formatResult(value: number): string {
	if (!Number.isFinite(value)) return "Error";
	if (Number.isNaN(value)) return "Error";
	if (value === 0) return "0";
	const abs = Math.abs(value);
	if (abs >= 1e15 || abs < 1e-9) {
		return value.toExponential(8).replace(/\.?0+e/, "e");
	}
	// Round to 12 significant digits to kill float fuzz, then trim.
	const r = Number.parseFloat(value.toPrecision(12));
	let s = String(r);
	if (s.includes(".") && !s.includes("e")) {
		s = s.replace(/(\.\d*?)0+$/, "$1").replace(/\.$/, "");
	}
	return s;
}

File: src-tauri/src/calc.rs
// Pretty Calc — Rust math engine.
// Tokenizer -> shunting-yard (RPN) -> evaluator. Used for committed `=` calls
// where the IPC cost is outweighed by exactness + raw f64 throughput.

use std::f64::consts::{E, PI};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AngleMode {
    Deg,
    Rad,
}

impl AngleMode {
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "DEG" | "deg" => Some(AngleMode::Deg),
            "RAD" | "rad" => Some(AngleMode::Rad),
            _ => None,
        }
    }

    pub fn to_rad(self, x: f64) -> f64 {
        match self {
            AngleMode::Deg => x * PI / 180.0,
            AngleMode::Rad => x,
        }
    }

    pub fn from_rad(self, x: f64) -> f64 {
        match self {
            AngleMode::Deg => x * 180.0 / PI,
            AngleMode::Rad => x,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
enum Token {
    Num(f64),
    Op(OpName),
    Func(String),
    ParenOpen,
    ParenClose,
    Postfix(PostfixOp),
    Comma,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum OpName {
    Add,
    Sub,
    Mul,
    Div,
    Pow,
    Mod,
    Neg,
    Implicit,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum PostfixOp {
    Fact,
    Percent,
}

impl OpName {
    fn precedence(self) -> u8 {
        match self {
            OpName::Pow | OpName::Neg => 5,
            OpName::Mul | OpName::Div | OpName::Mod | OpName::Implicit => 3,
            OpName::Add | OpName::Sub => 2,
        }
    }

    fn is_right_assoc(self) -> bool {
        matches!(self, OpName::Pow | OpName::Neg)
    }

    fn arity(self) -> usize {
        match self {
            OpName::Neg => 1,
            _ => 2,
        }
    }
}

const FUNCTIONS: &[&str] = &[
    "sin", "cos", "tan", "asin", "acos", "atan",
    "sinh", "cosh", "tanh", "asinh", "acosh", "atanh",
    "ln", "log", "log2", "sqrt", "cbrt", "exp",
    "abs", "floor", "ceil", "round", "sign",
];

fn is_function(name: &str) -> bool {
    FUNCTIONS.contains(&name)
}

fn is_ident_start(c: char) -> bool {
    c.is_ascii_alphabetic() || c == '_'
}

fn is_ident_cont(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

fn is_value_end(t: &Token) -> bool {
    matches!(t, Token::Num(_) | Token::ParenClose | Token::Postfix(_))
}

fn is_value_start(t: &Token) -> bool {
    matches!(t, Token::Num(_) | Token::Func(_) | Token::ParenOpen)
}

fn tokenize(input: &str) -> Result<Vec<Token>, CalcError> {
    let s: String = input.chars().filter(|c| !c.is_whitespace()).collect();
    let bytes: Vec<char> = s.chars().collect();
    let mut tokens: Vec<Token> = Vec::new();
    let mut i = 0;

    macro_rules! push_implicit {
        ($next:expr) => {
            if let Some(last) = tokens.last() {
                if is_value_end(last) && is_value_start(&$next) {
                    tokens.push(Token::Op(OpName::Implicit));
                }
            }
        };
    }

    while i < bytes.len() {
        let c = bytes[i];

        if c.is_ascii_digit() || c == '.' {
            let start = i;
            let mut dot = c == '.';
            i += 1;
            while i < bytes.len() {
                let d = bytes[i];
                if d.is_ascii_digit() {
                    i += 1;
                } else if d == '.' && !dot {
                    dot = true;
                    i += 1;
                } else {
                    break;
                }
            }
            let slice: String = bytes[start..i].iter().collect();
            let num: f64 = slice.parse().map_err(|_| CalcError::new("Malformed number"))?;
            push_implicit!(Token::Num(num));
            tokens.push(Token::Num(num));
            continue;
        }

        if is_ident_start(c) {
            let start = i;
            i += 1;
            while i < bytes.len() && is_ident_cont(bytes[i]) {
                i += 1;
            }
            let name: String = bytes[start..i].iter().collect::<String>().to_lowercase();
            match name.as_str() {
                "pi" => {
                    push_implicit!(Token::Num(PI));
                    tokens.push(Token::Num(PI));
                }
                "e" => {
                    push_implicit!(Token::Num(E));
                    tokens.push(Token::Num(E));
                }
                "ans" => return Err(CalcError::new("ans substituted by caller")),
                "mod" => tokens.push(Token::Op(OpName::Mod)),
                n if is_function(n) => {
                    push_implicit!(Token::Func(n.to_string()));
                    tokens.push(Token::Func(n.to_string()));
                }
                other => return Err(CalcError::new(&format!("Unknown name: {}", other))),
            }
            continue;
        }

        if c == '(' {
            if let Some(last) = tokens.last() {
                if is_value_end(last) {
                    tokens.push(Token::Op(OpName::Implicit));
                }
            }
            tokens.push(Token::ParenOpen);
            i += 1;
            continue;
        }
        if c == ')' {
            tokens.push(Token::ParenClose);
            i += 1;
            continue;
        }
        if c == ',' {
            tokens.push(Token::Comma);
            i += 1;
            continue;
        }
        if c == '!' {
            tokens.push(Token::Postfix(PostfixOp::Fact));
            i += 1;
            continue;
        }
        if c == '%' {
            tokens.push(Token::Postfix(PostfixOp::Percent));
            i += 1;
            continue;
        }
        if c == '+' || c == '-' {
            let is_unary = match tokens.last() {
                None => true,
                Some(t) => matches!(t, Token::Op(_) | Token::Func(_) | Token::Comma | Token::ParenOpen),
            };
            if is_unary {
                if c == '-' {
                    tokens.push(Token::Op(OpName::Neg));
                }
            } else {
                tokens.push(Token::Op(if c == '+' { OpName::Add } else { OpName::Sub }));
            }
            i += 1;
            continue;
        }
        if c == '*' || c == '\u{2217}' || c == '×' || c == '·' {
            tokens.push(Token::Op(OpName::Mul));
            i += 1;
            continue;
        }
        if c == '/' || c == '÷' {
            tokens.push(Token::Op(OpName::Div));
            i += 1;
            continue;
        }
        if c == '^' {
            tokens.push(Token::Op(OpName::Pow));
            i += 1;
            continue;
        }

        return Err(CalcError::new(&format!("Unexpected character: {}", c)));
    }

    Ok(tokens)
}

fn to_rpn(tokens: Vec<Token>) -> Result<Vec<Token>, CalcError> {
    let mut output: Vec<Token> = Vec::new();
    let mut stack: Vec<Token> = Vec::new();

    for t in tokens {
        match &t {
            Token::Num(_) => output.push(t),
            Token::Func(_) => stack.push(t),
            Token::Comma => {
                loop {
                    match stack.last() {
                        Some(Token::ParenOpen) => break,
                        Some(_) => output.push(stack.pop().unwrap()),
                        None => return Err(CalcError::new("Misplaced comma")),
                    }
                }
            }
            Token::Op(op) => {
                let cur = *op;
                while let Some(top) = stack.last() {
                    match top {
                        Token::Func(_) => {
                            output.push(stack.pop().unwrap());
                            continue;
                        }
                        Token::Op(top_op) => {
                            let top_prec = top_op.precedence();
                            let cur_prec = cur.precedence();
                            let right = cur.is_right_assoc();
                            if (right && top_prec > cur_prec) || (!right && top_prec >= cur_prec) {
                                output.push(stack.pop().unwrap());
                                continue;
                            }
                        }
                        _ => {}
                    }
                    break;
                }
                stack.push(Token::Op(cur));
            }
            Token::ParenOpen => stack.push(Token::ParenOpen),
            Token::ParenClose => {
                while let Some(top) = stack.pop() {
                    match top {
                        Token::ParenOpen => break,
                        x => output.push(x),
                    }
                }
                if let Some(Token::Func(_)) = stack.last() {
                    output.push(stack.pop().unwrap());
                }
            }
            Token::Postfix(_) => output.push(t),
        }
    }

    while let Some(t) = stack.pop() {
        if matches!(t, Token::ParenOpen | Token::ParenClose) {
            return Err(CalcError::new("Mismatched parentheses"));
        }
        output.push(t);
    }

    Ok(output)
}

fn factorial(n: f64) -> Result<f64, CalcError> {
    if n < 0.0 || n.fract() != 0.0 {
        return Err(CalcError::new("Factorial requires non-negative integer"));
    }
    if n > 170.0 {
        return Err(CalcError::new("Factorial too large"));
    }
    let mut r = 1.0;
    let k = n as u64;
    for i in 2..=k {
        r *= i as f64;
    }
    Ok(r)
}

fn apply_func(name: &str, x: f64, angle: AngleMode) -> Result<f64, CalcError> {
    let r = match name {
        "sin" => angle.to_rad(x).sin(),
        "cos" => angle.to_rad(x).cos(),
        "tan" => angle.to_rad(x).tan(),
        "asin" => angle.from_rad(x.asin()),
        "acos" => angle.from_rad(x.acos()),
        "atan" => angle.from_rad(x.atan()),
        "sinh" => x.sinh(),
        "cosh" => x.cosh(),
        "tanh" => x.tanh(),
        "asinh" => x.asinh(),
        "acosh" => x.acosh(),
        "atanh" => x.atanh(),
        "ln" => x.ln(),
        "log" => x.log10(),
        "log2" => x.log2(),
        "sqrt" => x.sqrt(),
        "cbrt" => x.cbrt(),
        "exp" => x.exp(),
        "abs" => x.abs(),
        "floor" => x.floor(),
        "ceil" => x.ceil(),
        "round" => x.round(),
        "sign" => x.signum(),
        _ => return Err(CalcError::new(&format!("Unknown function: {}", name))),
    };
    Ok(r)
}

fn evaluate_rpn(rpn: Vec<Token>, angle: AngleMode) -> Result<f64, CalcError> {
    let mut stack: Vec<f64> = Vec::new();

    for t in rpn {
        match t {
            Token::Num(n) => stack.push(n),
            Token::Postfix(op) => {
                let a = stack.pop().ok_or_else(|| CalcError::new("Syntax error"))?;
                let r = match op {
                    PostfixOp::Fact => factorial(a)?,
                    PostfixOp::Percent => a / 100.0,
                };
                stack.push(r);
            }
            Token::Op(op) => {
                let arity = op.arity();
                if arity == 1 {
                    let a = stack.pop().ok_or_else(|| CalcError::new("Syntax error"))?;
                    let r = match op {
                        OpName::Neg => -a,
                        _ => unreachable!("unary arity mismatch"),
                    };
                    stack.push(r);
                } else {
                    let b = stack.pop().ok_or_else(|| CalcError::new("Syntax error"))?;
                    let a = stack.pop().ok_or_else(|| CalcError::new("Syntax error"))?;
                    let r = match op {
                        OpName::Add => a + b,
                        OpName::Sub => a - b,
                        OpName::Mul | OpName::Implicit => a * b,
                        OpName::Div => {
                            if b == 0.0 {
                                return Err(CalcError::new("Division by zero"));
                            }
                            a / b
                        }
                        OpName::Pow => a.powf(b),
                        OpName::Mod => {
                            if b == 0.0 {
                                return Err(CalcError::new("Mod by zero"));
                            }
                            ((a % b) + b) % b
                        }
                        _ => unreachable!("binary arity mismatch"),
                    };
                    stack.push(r);
                }
            }
            Token::Func(name) => {
                let a = stack.pop().ok_or_else(|| CalcError::new(&format!("Missing argument for {}", name)))?;
                stack.push(apply_func(&name, a, angle)?);
            }
            _ => return Err(CalcError::new("Malformed expression")),
        }
    }

    if stack.len() != 1 {
        return Err(CalcError::new(if stack.is_empty() {
            "Empty input"
        } else {
            "Syntax error"
        }));
    }
    Ok(stack.pop().unwrap())
}

#[derive(Debug)]
pub struct CalcError {
    msg: String,
}

impl CalcError {
    pub fn new(msg: &str) -> Self {
        Self { msg: msg.to_string() }
    }
    pub fn message(&self) -> &str {
        &self.msg
    }
}

fn substitute_ans(input: &str, ans: f64) -> String {
    let mut out = String::with_capacity(input.len());
    let chars: Vec<char> = input.chars().collect();
    let mut i = 0;
    while i < chars.len() {
        let c = chars[i];
        if c.is_ascii_alphabetic() || c == '_' {
            let start = i;
            while i < chars.len() && (chars[i].is_ascii_alphanumeric() || chars[i] == '_') {
                i += 1;
            }
            let word: String = chars[start..i].iter().collect();
            if word.eq_ignore_ascii_case("ans") {
                out.push('(');
                out.push_str(&format_number(ans));
                out.push(')');
            } else {
                out.push_str(&word);
            }
        } else {
            out.push(c);
            i += 1;
        }
    }
    out
}

pub fn format_number(value: f64) -> String {
    if !value.is_finite() || value.is_nan() {
        return "Error".to_string();
    }
    if value == 0.0 {
        return "0".to_string();
    }
    let abs = value.abs();
    if abs >= 1e15 || abs < 1e-9 {
        return format!("{:e}", value);
    }
    let r = format!("{:.12}", value);
    let r = r.trim_end_matches('0').trim_end_matches('.');
    if r.is_empty() || r == "-" {
        "0".to_string()
    } else {
        r.to_string()
    }
}

pub fn evaluate(input: &str, angle: AngleMode, ans: f64) -> Result<f64, CalcError> {
    let with_ans = substitute_ans(input, ans);
    let trimmed = with_ans.trim();
    if trimmed.is_empty() {
        return Err(CalcError::new("Empty input"));
    }
    let tokens = tokenize(trimmed)?;
    let rpn = to_rpn(tokens)?;
    evaluate_rpn(rpn, angle)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn eval(s: &str) -> f64 {
        evaluate(s, AngleMode::Deg, 0.0).unwrap()
    }

    #[test]
    fn basic() {
        assert!((eval("1+2") - 3.0).abs() < 1e-12);
        assert!((eval("2*3+4") - 10.0).abs() < 1e-12);
        assert!((eval("2+3*4") - 14.0).abs() < 1e-12);
        assert!((eval("(2+3)*4") - 20.0).abs() < 1e-12);
        assert!((eval("2^3") - 8.0).abs() < 1e-12);
        assert!((eval("10/2/5") - 1.0).abs() < 1e-12);
    }

    #[test]
    fn unary_and_implicit() {
        assert!((eval("-2+3") - 1.0).abs() < 1e-12);
        assert!((eval("2(3)") - 6.0).abs() < 1e-12);
        assert!((eval("2pi") - 2.0 * PI).abs() < 1e-12);
    }

    #[test]
    fn funcs() {
        assert!((eval("sqrt(16)") - 4.0).abs() < 1e-12);
        assert!((eval("sin(30)") - 0.5).abs() < 1e-12);
        assert!((eval("ln(e)") - 1.0).abs() < 1e-12);
        assert!((eval("log(100)") - 2.0).abs() < 1e-12);
        assert!((eval("5!") - 120.0).abs() < 1e-9);
        assert!((eval("50%") - 0.5).abs() < 1e-12);
    }
}